File: apps/workers/src/Buteco.Workers/Knowledge/Entities/KnowledgeDocument.cs
namespace Buteco.Workers.Knowledge.Entities;

/// <summary>
/// Espelho de leitura de
/// <c>Buteco.Api.KnowledgeDocuments.Entities.KnowledgeDocument</c>.
/// </summary>
public class KnowledgeDocument
{
    public Guid Id { get; private set; }

    public Guid KnowledgeBaseId { get; private set; }

    public string Title { get; private set; } = null!;

    public string SourceType { get; private set; } = null!;

    public string ExtractedText { get; private set; } = null!;

    /// <summary>
    /// Coluna gerada pelo Postgres — nunca escrita por nenhum dos dois apps.
    /// </summary>
    public int ContentLengthBytes { get; private set; }

    public KnowledgeIndexingStatus IndexingStatus { get; private set; }

    public DateTimeOffset? IndexedAt { get; private set; }

    public string? FailureReason { get; private set; }

    /// <summary>
    /// Lida pelo consumidor de indexação da etapa seguinte, que grava
    /// condicionado a ela ainda ser a corrente.
    /// </summary>
    public int ContentRevision { get; private set; }

    /// <summary>
    /// Nulo significa "nunca indexado sob a regra de ContentHash" — linha criada
    /// antes desta capability existir. Escrito por <c>apps/api</c>.
    /// </summary>
    public string? ContentHash { get; private set; }

    /// <summary>Escrito pelo consumidor de indexação deste app.</summary>
    public int FragmentCount { get; private set; }

    /// <inheritdoc cref="FragmentCount"/>
    public int IndexingAttempts { get; private set; }

    /// <inheritdoc cref="FragmentCount"/>
    public DateTimeOffset? LastAttemptAt { get; private set; }

    public DateTimeOffset CreatedAt { get; private set; }

    public DateTimeOffset UpdatedAt { get; private set; }

    private KnowledgeDocument()
    {
    }
}

File: apps/frontend/src/features/knowledge-bases/types/knowledgeDocument.ts
// Espelha KnowledgeDocumentSummaryResponse e KnowledgeDocumentResponse de
// apps/api. Documento pertence à feature `knowledge-bases` por conceito de
// domínio, não por origem da rota (convenção 7).
//
// `indexingStatus` é união de string literal porque o enum vai para o fio como
// STRING, nunca como ordinal: KnowledgeIndexingStatus tem
// [JsonConverter(typeof(JsonStringEnumConverter<...>))] e
// KnowledgeWireFormatTests inspeciona o JSON bruto para afirmar exatamente
// "Pending"/"Indexing"/"Indexed"/"Failed" (convenção 12). Decodificar por índice
// seria o defeito que aquela convenção nasceu para impedir.
export type KnowledgeIndexingStatus = 'Pending' | 'Indexing' | 'Indexed' | 'Failed';

// `contentHash` NÃO EXISTE em resposta nenhuma, e por isso não está aqui.
//
// A coluna existe em KnowledgeDocument e tem um propósito só — Update() não
// devolve o documento para Pending quando o conteúdo extraído é idêntico ao
// gravado —, mas nem KnowledgeDocumentResponse nem
// KnowledgeDocumentSummaryResponse a projetam, e KnowledgeWireFormatTests não a
// menciona. Conferido no fio, não deduzido.
//
// É o campo que alguém vai procurar ao implementar o aviso de atualização. Não
// está aqui porque não dá para buscá-lo: a comparação de conteúdo é local, do
// `extractedText` carregado contra o editado (design.md, D5).
export interface KnowledgeDocumentSummary {
  id: string;
  knowledgeBaseId: string;
  title: string;
  sourceType: string;
  contentLengthBytes: number;
  indexingStatus: KnowledgeIndexingStatus;
  // Nulo enquanto o documento nunca foi indexado com sucesso; preenchido a
  // partir daí e PRESERVADO por atualizações, reindexações e falhas. É este
  // campo, e não o estado, que governa a exibição da contagem de fragmentos —
  // ver utils/documentIndexing.ts.
  indexedAt: string | null;
  failureReason: string | null;
  contentRevision: number;
  fragmentCount: number;
  indexingAttempts: number;
  lastAttemptAt: string | null;
  createdAt: string;
  updatedAt: string;
}

// A listagem não traz `extractedText` de propósito (design.md D14 da etapa 1):
// uma base com 50 documentos de 80 KB devolveria 4 MB de texto que nenhuma tela
// usa. Quem precisa do conteúdo — o modal de atualizar — busca o documento
// inteiro por GET /{id}.
export interface KnowledgeDocument extends KnowledgeDocumentSummary {
  extractedText: string;
}

// `sourceType` é string aberta validada em runtime contra os extratores
// registrados em apps/api, não enum fechado. Hoje só existe `markdown`, e a tela
// o mostra num campo desabilitado para deixar visível que o campo existe e que
// não há escolha.
export interface CreateKnowledgeDocumentInput {
  title: string;
  sourceType: string;
  content: string;
}

export type UpdateKnowledgeDocumentInput = CreateKnowledgeDocumentInput;
